api: add support for custom CA certificates via settings

arnaldo2792 opened the pull request:
**Issue number:**
#1277 

**Description of changes:**

```
684705a1 api: add support for custom CA certificates via settings
```

I'm still working on:

- [x] Migration
- [x] README documentation for new API

**Testing done:**

- [x] Pull down host container images
- [x] Send custom CA certificate through user data and verified the store was updated
- [x] Run pod/task/container in ecs-1, aws-k8s-1.19, aws-k8s-1.20
- [x] Run pod in vmware-1.20
- [x] Test container registry with custom CA

With bootstrap  container:

```Dockerfile
FROM alpine
RUN apk add openssl
ADD ./generate-certificate ./csr.conf ./config.yml ./
ENTRYPOINT ["sh", "generate-certificate"]
```

`generate-certificate` as:

```sh
#!/bin/sh

set -xe

PERSISTENT_DIR=/.bottlerocket/rootfs/mnt/certificates

mkdir -p $PERSISTENT_DIR

# Mount certificates as a tmpfs filesystem
mount -t tmpfs none $PERSISTENT_DIR
cp config.yml $PERSISTENT_DIR

# Generate 3 certificates
for i in $(seq 3); do
  # Generate CA Certificate and Key
  openssl genrsa -out $PERSISTENT_DIR/ca-$i.key 2048
  openssl req -x509 -new -nodes -key $PERSISTENT_DIR/ca-$i.key \
    -subj "/CN=bottlerocket/C=US/L=WASHINGTON" -days 1825 -out $PERSISTENT_DIR/ca-$i.crt
done

# Randomly select the CA certificate
TARGET=ca-$(( $RANDOM % 3 + 1 ))
ln -s $PERSISTENT_DIR/$TARGET.crt $PERSISTENT_DIR/ca.crt
ln -s $PERSISTENT_DIR/$TARGET.key $PERSISTENT_DIR/ca.key

# Generate SSL/TLS Certificates
openssl genrsa -out $PERSISTENT_DIR/server.key 2048
openssl req -new -key $PERSISTENT_DIR/server.key -out $PERSISTENT_DIR/server.csr -config csr.conf
openssl x509 -req -in $PERSISTENT_DIR/server.csr -CA $PERSISTENT_DIR/ca.crt -CAkey $PERSISTENT_DIR/ca.key \
  -CAcreateserial -out $PERSISTENT_DIR/server.crt -days 10000 -extensions req_ext \
  -extfile csr.conf

# Save all certificates as a bundle
BUNDLE=$(cat $PERSISTENT_DIR/ca-1.crt $PERSISTENT_DIR/ca-2.crt $PERSISTENT_DIR/ca-3.crt | base64 -w0)
apiclient set pki.local-registry.data=$BUNDLE \
  pki.local-registry.trusted=true
```

Deployed local registry with pod spec (and the equivalent of an ECS task):

```yml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: ca-certificates
spec:
  serviceName: ca-certificates
  replicas: 1
  selector:
    matchLabels:
      app: ca-certificates
  template:
    metadata:
      labels:
        app: ca-certificates
    spec:
      volumes:
        - name: certificates
          hostPath:
            path: /mnt/certificates
            type: Directory
        - name: config
          hostPath:
            path: /mnt/certificates/config.yml
      containers:
        - name: registry
          image: docker.io/library/registry:2
          ports:
            - containerPort: 5000
              hostPort: 5000
          volumeMounts:
            - mountPath: /certificates
              name: certificates
              readOnly: true
            - mountPath: /etc/docker/registry/config.yml
              name: config
              readOnly: true
```

Test that I get a 404 instead of an error related to a cerficate:

```sh
bash-5.0# ctr -a /run/host-containerd/containerd.sock image pull localhost:5000/test:latest
INFO[0000] trying next host - response was http.StatusNotFound  host="localhost:5000"
ctr: failed to resolve reference "localhost:5000/test:latest": localhost:5000/test:latest: not found
# Set the bundle as distrusted
bash-5.0# apiclient set pki.local-registry.trusted=false
# Get error `certificate signed by unknown authority`
bash-5.0# ctr -a /run/host-containerd/containerd.sock image pull localhost:5000/test:latest
INFO[0000] trying next host                              error="failed to do request: Head \"https://localhost:5000/v2/test/manifests/latest\": x509: certificate signed by unknown authority" host="localhost:5000"
ctr: failed to resolve reference "localhost:5000/test:latest": failed to do request: Head "https://localhost:5000/v2/test/manifests/latest": x509: certificate signed by unknown authority
```
**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
api: add support for custom CA certificates via settings

This commit adds support to update the local trusted certificates store
through the API using custom CA certificates.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes migration

arnaldo2792 commented:
Forced push includes:

- Documentation
- Minor updates  in  README
- Fix in the service triggered when `pki` settings change

bcressey commented on README.md:
```diff
@@ -465,6 +465,33 @@ Here are the metrics settings:
+  pki.my-trusted-bundle.data="W3N..." \
+  pki.my-trusted-bundle.trusted=true  \
+  pki.dont-trust-these.data="N3W..."  \
+  pki.dont-trust-there.trusted=false
```
We don't provide very many `apiclient set` examples elsewhere. Are we leading with this approach because we expect the size of certificate bundles to exceed what's practical to pass through user-data? That might be worth a comment if so.

arnaldo2792 replied:
I didn't have any intention to lead users to use `apiclient`, but you made a great point and I'll document what could happen if users send bundles through user data.

bcressey commented on README.md:
```diff
@@ -516,7 +543,7 @@ There are a few important caveats to understand about host containers:
 
 Because of these caveats, host containers are only intended for special use cases.
-We use it for the control container because it needs to be available early to give you access to the OS, and we use it for the admin container because it needs high levels of privilege and because you need it to debug when orchestration isn't working.
+We use them for the control container because it needs to be available early to give you access to the OS, and for the admin container because it needs high levels of privilege and because you need it to debug when orchestration isn't working.
```
Seems like a good edit but might be better as a separate commit.

bcressey commented on README.md:
```diff
@@ -535,7 +562,7 @@ Bootstrap containers have access to the underlying root filesystem on `/.bottler
 
 Bootstrap containers are set up to run after the systemd `configured.target` unit is active.
-The containers' systemd unit depends on this target (and not on any of the bootstrap containers' peers) which means that bootstrap containers will not execute in a deterministic order
+The containers' systemd unit depends on this target (and not on any of the bootstrap containers' peers) which means that bootstrap containers will not execute in a deterministic order.
```
I'd roll this into the other "fixup docs" commit.

bcressey commented on sources/Cargo.lock:
```diff
@@ -1703,9 +1796,9 @@ checksum = "490cc448043f947bae3cbee9c203358d62dbee0db12107a74be5c30ccfd09771"
 [[package]]
 name = "memchr"
-version = "2.4.0"
+version = "2.3.4"
```
Lots of downgrades like this. Is this a rebase gone awry, or do the new dependencies force us into older versions?

webern replied:
Maybe copy from develop and overwrite the Cargo.lock, then build.

arnaldo2792 replied:
I did regenerate the Cargo.lock file, no bad rebase, as Ben said, the new dependencies forced  older versions :cry:

bcressey commented on sources/api/certdog/README.md:
```diff
@@ -0,0 +1,11 @@
+
+Current version: 0.1.0
+
+  certdog is a tool to manage the trusted certificates store. It adds/removes
```
```suggestion
  certdog is a tool to manage the trusted certificates store. It adds and removes
```

bcressey commented on sources/api/migration/migrations/v1.1.4/add-custom-certificates/src/main.rs:
```diff
@@ -0,0 +1,20 @@
+use migration_helpers::{migrate, Result};
+use std::process;
+
+/// We added the setting `pki`
```
nit:
```suggestion
/// We added settings and services for `pki`
```

bcressey commented on sources/models/src/aws-dev/mod.rs:
```diff
@@ -5,7 +5,7 @@ use std::collections::HashMap;
 use crate::{
     AwsSettings, BootstrapContainer, HostContainer, KernelSettings, MetricsSettings,
-    NetworkSettings, NtpSettings, UpdatesSettings,
+    NetworkSettings, NtpSettings, PEMCertificate, UpdatesSettings,
```
If NTP is "Ntp" and AWS is "Aws" then PEM should probably be "Pem":
```suggestion
    NetworkSettings, NtpSettings, PemCertificate, UpdatesSettings,
```

bcressey commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -83,6 +84,16 @@ pub mod error {
+
+        #[snafu(display("Invalid x509 certificate: {}", source))]
+        InvalidX509Certificate {
+            source: x509_parser::nom::Err<x509_parser::error::X509Error>,
```
Can this be simplified to `x509_parser::error::X509Error`?

arnaldo2792 replied:
I remember I tried, but I can double check.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -618,3 +619,65 @@ mod test_valid_container_mode {
+
+    #[test]
+    fn invalid_pem_certificate() {
+        assert!(PEMCertificateString::try_from(
```
Can we also test that it doesn't parse input without the BEGIN and END markers?

arnaldo2792 replied:
Not passing the markers will result in an empty iterator returned by `x509_parser::pem::Pem::iter_from_buffer`, so I think I can modify the code to do something like:

```rust
 let mut empty = true;
 for (_, pem) in x509_parser::pem::Pem::iter_from_buffer(&decoded_bytes).enumerate() {
            // Parse buffer into a PEM object, then to a x509 certificate
            let pem = pem.context(error::InvalidPEM)?;
            pem.parse_x509().context(error::InvalidX509Certificate)?;
            // Or skip the if and always set empty = false
            if empty {
                   empty = false;
            }
        }

  assert!(!empty)
```

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+
+use model::modeled_types::Identifier;
+
+// FIXME Get from configuration in the future
+const DEFAULT_API_SOCKET: &str = "/run/api.sock";
```
$1 to whoever who actually fixes this FIXME. 😢 

(obviously not a blocker here)

tjkirch replied:
Or just removes the FIXMEs; I'm not sure the idea makes sense any more.

webern replied:
A constants crate would at least consolidate the FIXME to one place, no?

arnaldo2792 replied:
I was planning to start the new crate in this PR, provided that I don't have to actually fix all the FIXMEs.

bcressey reviewed (commented):
Notes for everything except `certdog` which I need to review in more detail.

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+    trusted_store: String,
+    source_bundle: String,
+    comment: bool,
+}
```
I'd prefer using `argh` again here, as we did with `prairiedog`.

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+/// and distrusted PEM certificates
+async fn get_certificate_bundles<P>(
+    socket_path: P,
+) -> Result<(Vec<x509_parser::pem::Pem>, Vec<x509_parser::pem::Pem>)>
```
Maybe something like this instead, with a structured type?

```suggestion
) -> Result<CertBundle>

...

struct CertBundle {
    trusted_certs: Vec<x509_parser::pem::Pem>,
    distrusted_certs: Vec<x509_parser::pem::Pem>,
}

impl TryFrom<HashMap<Identifier, model::PEMCertificate>> for CertBundle {};
```

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+{
+    debug!("Querying the API for settings");
+
+    let method = "GET";
+    let uri = API_SETTINGS_URI;
+    let (_code, response_body) = apiclient::raw_request(&socket_path, uri, method, None)
+        .await
+        .context(error::APIRequest { method, uri })?;
```
Can we query for just the settings we need, instead of retrieving all of them?

arnaldo2792 replied:
I have to read the code in `apiclient` to check if it allows to retrieve specific keys

arnaldo2792 replied:
As far as I can tell,  no, [`raw_request`](https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/apiclient/src/lib.rs
) doesn't allow to filter specific keys.

tjkirch replied:
It's not a part of the function, it's the URI you send to.  There are parameters like [keys](https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/openapi.yaml#L21) and [prefix](https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/api/openapi.yaml#L32) that let you select what you want to get back.

arnaldo2792 replied:
Per our offline discussion, we will work on an improved helper to serialize/deserialize specific keys rather than the entire settings store.

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+    let trusted_store = trusted_store.as_ref();
+
+    // The default bundle includes the certificates shipped with the OS
+    let default_bundle = fs::File::open(source_bundle).context(error::ReadFile {
```
Let's have a more specific read error, to match the more specific write error.
```suggestion
    let default_bundle = fs::File::open(source_bundle).context(error::ReadSourceBundle {
```

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+        &trusted_store,
+        pems_to_string(&trusted_bundle, add_comment)?,
+    )
+    .context(error::UpdateTrustStore)?;
```
```suggestion
    .context(error::UpdateTrustedStore)?;
```

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+                })?
+            }
+
+            "-c" | "--comment" => final_args.comment = true,
```
Why not always add comments?

arnaldo2792 replied:
Comments aren't necessary for the final bundle, but I can make them part of it by default.

bcressey commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+
+/// Returns a string from the common name, organizational unit or organization
+/// fields in the certificate
+fn comment_for_pem(pem: &x509_parser::pem::Pem) -> Result<Option<String>> {
+    let cert = pem.parse_x509().context(error::ParseX509Certificate)?;
+    let subject = cert.tbs_certificate.subject;
+
+    if let Some(common_name) = subject.iter_common_name().next() {
+        if let Ok(common_name_str) = common_name.as_str() {
+            return Ok(Some(common_name_str.to_string()));
+        }
+    } else if let Some(organizational_unit) = subject.iter_organizational_unit().next() {
+        if let Ok(organizational_unit_str) = organizational_unit.as_str() {
+            return Ok(Some(organizational_unit_str.to_string()));
+        }
+    } else if let Some(organization) = subject.iter_organization().next() {
+        if let Ok(organization_str) = organization.as_str() {
+            return Ok(Some(organization_str.to_string()));
+        }
+    }
+
+    Ok(None)
+}
```
Is this what's required to mimic the default comments in our CA certificates bundle? It seems like a lot of work to produce output that might not be satisfactory in all cases.

```suggestion
fn comment_for_pem(pem: &x509_parser::pem::Pem) -> Result<Option<String>> {
    let cert = pem.parse_x509().context(error::ParseX509Certificate)?;
    let subject = cert.tbs_certificate.subject;
    let comment = subject                                                                                                                              
        .iter_common_name()                                                                                                                            
        .chain(subject.iter_organizational_unit())                                                                                                     
        .chain(subject.iter_organization())                                                                                                            
        .next();                                                                                                                                       
    Ok(comment                                                                                                                                         
        .and_then(|c| c.as_str().ok())                                                                                                                 
        .and_then(|c| Some(c.to_string())))
}
```

arnaldo2792 replied:
Yes, it is to mimic that. I haven't nail the logic that the `trust` binary uses, but this is close to what the binary produces.

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+}
+
+/// Returns a tuple with two lists, for trusted and distrusted certificates
+fn split_bundles(
```
This is a pure function (nice!) and should have unit tests.

(I meant to bundle this with a review but hit the wrong button)

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+
+/// Updates the trusted certificates store, removing the distrusted certificates
+/// from the final bundle
+fn update_trusted_store<P>(
```
Using `TempFile::TempDir` and a checked-in bundle, I believe this could also be unit tested.

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+}
+
+/// Transforms a PEM object into a PEM formatted string
+fn pem_to_string(pem: &x509_parser::pem::Pem, add_comment: bool) -> Result<String> {
```
This and `pems_to_string` also seem test-able, but would be covered by `update_trusted_store`.

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+
+    Global arguments:
+        [ --socket-path PATH ]
+        [ --source-bundle PATH]
+        [ --trusted-store PATH]
+        [ --comment]
```
nit

```suggestion
        [ --source-bundle PATH ]
        [ --trusted-store PATH ]
        [ --comment ]
```

webern commented on sources/Cargo.toml:
```diff
@@ -37,6 +38,7 @@ members = [
     "api/migration/migrations/v1.1.2/control-container-v0-5-1",
     "api/migration/migrations/v1.1.3/kubelet-cpu-manager-state",
     "api/migration/migrations/v1.1.3/kubelet-cpu-manager",
+    "api/migration/migrations/v1.1.4/add-custom-certificates",
```
Looks like the version here doesn't match `Release.toml`.

```suggestion
    "api/migration/migrations/v1.2.0/add-custom-certificates",
```

webern commented on sources/models/fixtures/test-pem:
```diff
@@ -0,0 +1 @@
+LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURVekNDQWp1Z0F3SUJBZ0lVVUNBbXU1MWU1NDZIakpGVEd6Z1BpS2ZxZXVVd0RRWUpLb1pJaHZjTkFRRUwKQlFBd09URVZNQk1HQTFVRUF3d01ZbTkwZEd4bGNtOWphMlYwTVFzd0NRWURWUVFHRXdKVlV6RVRNQkVHQTFVRQpCd3dLVjBGVFNFbE9SMVJQVGpBZUZ3MHlNVEEzTURjeU1UQXhNekJhRncweU5qQTNNRFl5TVRBeE16QmFNRGt4CkZUQVRCZ05WQkFNTURHSnZkSFJzWlhKdlkydGxkREVMTUFrR0ExVUVCaE1DVlZNeEV6QVJCZ05WQkFjTUNsZEIKVTBoSlRrZFVUMDR ...
```
Is `fixtures` a canonical name for a test data directory? In migrator I used `tests/data` because `tests` is the Rust built-in directory for 'integration' tests. I wonder how many variations of directory names we have for test input data.

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+            source: apiclient::Error,
+        },
+
+        #[snafu(display("Unable to decode base64 from certificate '{}': '{}'", name, source))]
```
nit

```suggestion
        #[snafu(display("Unable to decode base64 from certificate '{}': {}", name, source))]
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+            source: x509_parser::nom::Err<x509_parser::error::X509Error>,
+        },
+
+        #[snafu(display("Error while reading file {}: '{}'", path.display(), source))]
```
nit

```suggestion
        #[snafu(display("Error while reading file '{}': {}", path.display(), source))]
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+        ReadFile { path: PathBuf, source: io::Error },
+
+        #[snafu(display(
+            "Error deserializing response as JSON from {} to {}: {}",
```
nit:

(or maybe it could be better than my edit, but slightly awkward as is)

```suggestion
            "Error deserializing '{}' response '{}': {}",
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,401 @@
+        #[snafu(display("Failed to update trust store: {}", source))]
+        UpdateTrustStore { source: io::Error },
+
+        #[snafu(display("failed to write to pem string: {}", source))]
```
nit

```suggestion
        #[snafu(display("Failed to write to pem string: {}", source))]
```

webern commented on sources/models/shared-defaults/defaults.toml:
```diff
@@ -116,3 +116,12 @@ restart-commands = ["/usr/bin/bootstrap-containers create-containers"]
 [metadata.settings.bootstrap-containers]
 affected-services = ["bootstrap-containers"]
+
+# Certdog
```
Are there services that need to be restarted when the bundle changes? Or does everything read the bundle 'just in time'?

arnaldo2792 replied:
I think everything access to the bundle at runtime. In my test, I didn't have to restart `containerd` when I updated the bundle.

webern commented on sources/models/src/modeled_types/mod.rs:
```diff
@@ -9,6 +9,7 @@
 pub mod error {
     use regex::Regex;
     use snafu::Snafu;
+    use x509_parser;
```
nit, I think this is unused because the symbols are fully-qualified below?

```suggestion
```

arnaldo2792 commented:
Forced push includes:

- Updated README
- Moved nits in README unrelated to CA certificates, to their own commit
- Use `argh`  instead of custom arguments parser
- Fixed nit in certdog's documentation
- Always use certificate comments
- Added `CertBundle` struct instead of tuple of `Vec`s
- Changed error `ReadFile` to `ReadSourceBundle`
- Refactor `comment_for_pem` as suggested

arnaldo2792 commented:
Forced push includes:
- Fix migration version  from (1.1.3, 1.2.0) to (1.1.4, 1.2.0)
- Rename migration to 1.2.0
- Unit tests in certdog for `split_bundles` and  `update_trusted_store`
- Various nits in error messages
- Move test data from `fixtures` to `tests/data`
- Simplified `comment_for_pem`
- Validation to check if the bundle's data doesn't contain any valid PEM certificate
- Additional unit tests for `PemCertificateString`

bcressey reviewed (commented):
The docs fix commit is showing up as empty.

webern commented on sources/api/certdog/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+tempfile = "3.2.0"
+
+[build-dependencies]
+cargo-readme = "3.1"
```
Test-only dependencies can go in `dev-dependencies` and they are not compiled during non-test builds.

```suggestion
x509-parser = "0.9.2"

[build-dependencies]
cargo-readme = "3.1"

[dev-dependencies]
tempfile = "3.2.0"
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+            source: x509_parser::nom::Err<x509_parser::error::X509Error>,
+        },
+
+        #[snafu(display("Error while reading bundle from file {}: '{}'", path.display(), source))]
```
Nit: we always have the 'thing' in single quotes and the error at the end without quotes.

```suggestion
        #[snafu(display("Error while reading bundle from file '{}': {}", path.display(), source))]
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+        #[snafu(display("Error while reading bundle from file {}: '{}'", path.display(), source))]
+        ReadSourceBundle { path: PathBuf, source: io::Error },
+
+        #[snafu(display("Error deserializing response from '{}': '{}'", uri, source))]
```
Nit: we always have the 'thing' in single quotes and the error at the end without quotes.

```suggestion
        #[snafu(display("Error deserializing response from '{}': {}", uri, source))]
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+        UpdateTrustedStore { source: io::Error },
+
+        #[snafu(display("Failed to write to pem string: {}", source))]
+        WritePEMString { source: std::fmt::Error },
```
Nit: per Rust style and consistent with elsewhere, Rust acronyms are given as a CamelCased word.

```suggestion
        WritePemString { source: std::fmt::Error },
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+            source: apiclient::Error,
+        },
+
+        #[snafu(display("Unable to decode base64 from certificate '{}': '{}'", name, source))]
```
Nit: we always have the 'thing' in single quotes and the error at the end without quotes.

```suggestion
        #[snafu(display("Unable to decode base64 from certificate '{}': {}", name, source))]
```

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+
+#[cfg(test)]
+mod test_certdog {
+    use model;
+    use model::modeled_types::{Identifier, PemCertificateString};
```
Unless there's a name issue I don't know about, probably cleaner to import `PemCertificate` and drop the import of `model`.

```suggestion
    use model::modeled_types::{Identifier, PemCertificate, PemCertificateString};
```

arnaldo2792 replied:
PemCertificate isn't a modeled type, it is a model, and in other crates, we don't include `model` for some reason, we  always do `model::TheModel`

webern replied:
OK. 🤷

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+        let mut bundle = HashMap::new();
+        bundle.insert(
+            Identifier::try_from("trusted").unwrap(),
+            model::PemCertificate {
```
```suggestion
            PemCertificate {
```

arnaldo2792 replied:
Same as above, we don't include `model` in other crates

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+        );
+        bundle.insert(
+            Identifier::try_from("distrusted").unwrap(),
+            model::PemCertificate {
```
```suggestion
            PemCertificate {
```

arnaldo2792 replied:
Same

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+        );
+        bundle.insert(
+            Identifier::try_from("distrusted-without-flag").unwrap(),
+            model::PemCertificate {
```
```suggestion
            PemCertificate {
```

arnaldo2792 replied:
Same

webern commented on sources/api/certdog/src/main.rs:
```diff
@@ -0,0 +1,355 @@
+            trusted_certs,
+            distrusted_certs: Vec::new(),
+        };
+        assert!(super::update_trusted_store(certs_bundle, &trusted_store, &source_bundle).is_ok());
```
It's common to add `use super::*` at the top of `mod test_certdog`

```suggestion
        assert!(update_trusted_store(certs_bundle, &trusted_store, &source_bundle).is_ok());
```

webern reviewed (approved):
Thank you for adding unit tests! 🥼 

Just non-blocking cleanup nits here. Nice job.

arnaldo2792 committed:
docs: minor fix in host containers documentation

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

arnaldo2792 commented:
Forced push includes:

- Fixed empty commit
- Changed migrations to 1.1.5 instead of 1.2.0
- Cargo.toml includes `dev-dependencies` for certdog
- Nits in error names and messages

arnaldo2792 commented:
Forced push moves migrations back to 1.2.0
